flow/diff_context.h

```diff
@@ -0,0 +1,134 @@
+#ifndef FLUTTER_FLOW_DIFF_CONTEXT_H_
+#define FLUTTER_FLOW_DIFF_CONTEXT_H_
+
+#include <vector>
+#include "flutter/fml/macros.h"
+#include "third_party/skia/include/core/SkMatrix.h"
+#include "third_party/skia/include/core/SkRect.h"
+
+namespace flutter {
+
+// Represents a region on screen to which layer subtree has painted anything
+class PaintRegion {
+ public:
+  PaintRegion() {}
+  PaintRegion(std::shared_ptr<std::vector<SkRect>> rects,
+              size_t from,
+              size_t to,
+              bool has_readback)
+      : rects_(rects), from_(from), to_(to), has_readback_(has_readback) {}
+
+  std::vector<SkRect>::const_iterator begin() const {
+    assert(is_valid());
+    return rects_->begin() + from_;
+  }
+
+  std::vector<SkRect>::const_iterator end() const {
+    assert(is_valid());
+    return rects_->begin() + to_;
+  }
+
+  SkRect GetBounds() const;
+
+  bool is_valid() const { return rects_ != nullptr; }
+
+  bool has_readback() const { return has_readback_; }
+
+ private:
+  std::shared_ptr<std::vector<SkRect>> rects_;
+  size_t from_ = 0;
+  size_t to_ = 0;
+  bool has_readback_ = false;
+};
+
+class DiffContext {
+ public:
+  explicit DiffContext(double device_pixel_aspect_ratio);
+
+  class SubtreeHandle;
+
+  // Starts a new subtree. Subtree is ended when handle gets out of scope;
+  // All modifications (transform, cull rect) will be restored
+  [[nodiscard]] SubtreeHandle BeginSubtree();
+
+  // Pushes additional transform for current subtree
+  void PushTransform(const SkMatrix& transform);
+
+  // Pushes cull rect for current subtree
+  bool PushCullRect(const SkRect& clip);
+
+  SkMatrix GetTransform() const { return state_.transform; }
+
+  SkRect GetCullRect() const { return state_.cull_rect; }
+
+  // Sets the dirty flag on current subtree;
+  // When the subtree is closed, entire paint region of the subtree will be
+  // added to damage; Additionally previous_paint_region, which is supposed
+  // to be paint region of previous subtree, will also be added to damage.
+  void MarkSubtreeDirty(
+      const PaintRegion& previous_paint_region = PaintRegion());
+
+  bool IsSubtreeDirty() const { return state_.dirty; }
+
+  // Add paint region for layer; rect must be in "local" (layer) coordinates
+  void AddPaintRegion(const SkRect& rect);
+
+  // Add entire paint region for current subtree
+  void AddPaintRegion(const PaintRegion& region);
```

Comment:
Well, right now we do only one bounds rect, but it's possible that in future you will want more granularity (i.e. it's not ideal if two spinners in opposed corners cause redraw of entire screen). More importantly, once this is settled, I'd like to implement raster cache for backdrop filter. DiffContext would be used to detect whether the backdrop snapshot needs to be invalidated, and for this using bounds rect for this would potentially lead to many unnecessary invalidations.

Reply:
Also, the vector of rects is shared between all paint regions. So each paint region is just one shared_ptr and two indices.